### ISSUE TITLE
iwyu removes include when declaration is only used in uninstantiated template function

### ISSUE THREAD

**Issue by tberghammer** [include-what-you-use/include-what-you-use#965]
```
// a.h
struct Foo {};

// b.h
#include "a.h"
template <typename T> int size() { return sizeof(T); }
template <int X> int foo() { return X + size<Foo>(); }

// b.cc
#include "b.h"

// c.cc
#include "b.h"
int bar() { return foo<5>(); }
```

```
./include-what-you-use b.cc

b.h should add these lines:
struct Foo;

b.h should remove these lines:
- #include "a.h"  // lines 1-1

The full include-list for b.h:
struct Foo;
---

(b.cc has correct #includes/fwd-decls)
```

IWYU decides that we don't need the full definition of Foo in b.h because we never instantiated size<Foo>() but it means that when trying to compile c.cc it will fail to compile as we won't have the full definition of Foo.

P.S.: size<T>() is a simplification on std::make_unique<T> what fails the same way as the one above

**Comment by kimgr** [comment]
Thanks, that does look like something IWYU should be able to catch (try `-Xiwyu -v7` to increase verbosity level, and you'll see that it's traversing the uninstantiated template, but fails to report the full use of the template argument type for `sizeof`).

**Comment by tberghammer** [comment]
Is your expectation is that when we traverse the uninstantiated template of `size` then we will report a full use of `T` and then when we hit `size<Foo>` we look up the fact that `size<T>` will require the full use of `T` and based on that require a full use of `Foo` as well?

At the moment the first step (reporting a full use of `T` inside `size`) doesn't happen but I expect that to be something fairly easy to fix.

**Comment by kimgr** [comment]
Yes, something like that. The template handling in IWYU is complex enough that I've never gotten my head around it completely, but it seems to me if we see a template instantiation (even inside another uninstantiated template), we should be able to act on it.

**Comment by bolshakov-a** [comment]
In fact, compilers accept a forward-declaration of `Foo` for `foo()` definition although it is a non-dependent part of `foo()` template body: https://godbolt.org/z/an4T9xa6M. So I think there should be an author-intent policy here.

The problem is that IWYU doesn't report `Foo` in `c.cc` even when `Foo` is forward-declared in `b.h` instead of including, and there is a direct `#include "a.h"` in `c.cc` (so IWYU wrongly suggests to remove that `#include`).
